src/Functions/FunctionDynamicAdaptor.h

```diff
@@ -0,0 +1,79 @@
+    {
+    }
+
+    String getName() const override { return function_overload_resolver->getName(); }
+
+protected:
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const final;
+    ColumnPtr executeDryRunImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const final;
+
+    /// Disable all default implementations, they will be used later.
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool useDefaultImplementationForNothing() const override { return false; }
+    bool useDefaultImplementationForConstants() const override { return false; }
+    bool useDefaultImplementationForLowCardinalityColumns() const override { return false; }
+    bool useDefaultImplementationForSparseColumns() const override { return false; }
+    bool canBeExecutedOnDefaultArguments() const override { return false; }
+
+private:
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count, bool dry_run) const;
+
+    /// We remember the original IFunctionOverloadResolver to be able to build function for types inside Dynamic column.
+    std::shared_ptr<const IFunctionOverloadResolver> function_overload_resolver;
+    size_t dynamic_argument_index;
+};
+
+class FunctionBaseDynamicAdaptor final : public IFunctionBase
+{
+public:
+    FunctionBaseDynamicAdaptor(std::shared_ptr<const IFunctionOverloadResolver> function_overload_resolver_, DataTypes arguments_);
+
+    String getName() const override { return function_overload_resolver->getName(); }
+
+    const DataTypes & getArgumentTypes() const override { return arguments; }
+    const DataTypePtr & getResultType() const override { return return_type; }
+
+    ExecutableFunctionPtr prepare(const ColumnsWithTypeAndName & /*arguments*/) const override
+    {
+        return std::make_unique<ExecutableFunctionDynamicAdaptor>(function_overload_resolver, dynamic_argument_index);
+    }
+
+    bool isSuitableForConstantFolding() const override { return false; }
+
+    bool isStateful() const override { return function_overload_resolver->isStateful(); }
+
+    bool isDeterministic() const override { return function_overload_resolver->isDeterministic(); }
+
+    bool isDeterministicInScopeOfQuery() const override { return function_overload_resolver->isDeterministicInScopeOfQuery(); }
+
+    bool isServerConstant() const override { return function_overload_resolver->isServerConstant(); }
+
+    bool isShortCircuit(ShortCircuitSettings & settings, size_t number_of_arguments) const override { return function_overload_resolver->isShortCircuit(settings, number_of_arguments); }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo &) const override { return true; }
```

Comment:
It doesn't matter, because we want to execute function with Dynamic argument always lazily in short circuit functions, because:
- Executing on Dynamic may be not so efficient and lazy executiong after filtering can be more efficient.
- Executiong function with Dynamic argument may throw an exception (for example because of incimpatible argument type for some of the variants).

Simple example when underlying function is marked as not suitable for lazy execution, but with Dynamic argument we should execute it lazily:
```sq;
create table test (d Dynamic) engine=Memory;
insert into test values (1::UInt8), (2::UInt16), (3::UInt32), (4::UInt64), ('Hello');
```

```sql
:) select * from test where dynamicType(d) != 'String' and d % 2 = 0

   ┌─d─┐
1. │ 2 │
2. │ 4 │
   └───┘
```

Here if `modulo` function wasn't lazy executed, we would get exception `Illegal types String and UInt8 of arguments of function modulo`

Reply:
You are right, I always misunderstand the goal of `isSuitableForShortCircuitArgumentsExecution`. Thanks for the explanation though!